File: src_langgraph/memory.py
from __future__ import annotations

from typing import Any

from src_shared.contracts import WorkflowRunRequest, WorkflowRunResponse

DEFAULT_MEMORY_CONTEXT_KEY = "dialog_memory"
DEFAULT_ARTIFACT_CONTEXT_KEY = "artifact_memory"


def memory_from_context(context: dict[str, Any] | None) -> tuple[str, list[dict[str, str]], dict[str, Any]]:
    base = context if isinstance(context, dict) else {}
    extra = {k: v for k, v in base.items() if k != DEFAULT_MEMORY_CONTEXT_KEY}
    payload = base.get(DEFAULT_MEMORY_CONTEXT_KEY)
    if not isinstance(payload, dict):
        payload = {}
    summary = str(payload.get("summary") or "").strip()
    recent = _normalize_turns(payload.get("recent_turns"))
    return summary, recent, extra


def context_with_memory(*, summary: str, recent_turns: list[dict[str, str]], extra: dict[str, Any]) -> dict[str, Any]:
    out = dict(extra or {})
    out[DEFAULT_MEMORY_CONTEXT_KEY] = {
        "summary": str(summary or "").strip(),
        "recent_turns": _normalize_turns(recent_turns),
    }
    return out


def apply_edit_rewrite(
    *,
    summary: str,
    recent_turns: list[dict[str, str]],
    edit_turn_id: str | None,
    recent_messages_limit: int,
    summary_max_chars: int,
) -> tuple[str, list[dict[str, str]]]:
    if not edit_turn_id:
        return compact_memory(summary=summary, recent_turns=recent_turns, recent_messages_limit=recent_messages_limit, summary_max_chars=summary_max_chars)

    idx = -1
    for i, turn in enumerate(recent_turns):
        if turn.get("role") == "user" and turn.get("turn_id") == edit_turn_id:
            idx = i
            break
    if idx < 0:
        return "", []

    # Remove edited user turn and everything after it, then rebuild compact memory.
    rewritten = recent_turns[:idx]
    return compact_memory(summary="", recent_turns=rewritten, recent_messages_limit=recent_messages_limit, summary_max_chars=summary_max_chars)


def compact_memory(
    *,
    summary: str,
    recent_turns: list[dict[str, str]],
    recent_messages_limit: int,
    summary_max_chars: int,
) -> tuple[str, list[dict[str, str]]]:
    normalized = _normalize_turns(recent_turns)
    safe_limit = max(2, int(recent_messages_limit))
    safe_summary_max = max(200, int(summary_max_chars))
    if len(normalized) <= safe_limit:
        return _trim_tail(summary, safe_summary_max), normalized
    overflow = normalized[:-safe_limit]
    kept = normalized[-safe_limit:]
    next_summary = _merge_summary(summary, overflow, safe_summary_max)
    return next_summary, kept


def build_dialog_context(
    *,
    summary: str,
    recent_turns: list[dict[str, str]],
    extra: dict[str, Any] | None,
    max_chars: int,
) -> str:
    sections: list[str] = []
    artifact_block = _artifact_context_block(extra)
    if artifact_block:
        sections.append(artifact_block)
    if summary.strip():
        sections.append(f"Summary:\n{summary.strip()}")
    if recent_turns:
        lines = [
            f"- {_role_label(turn['role'])}: {turn['text']}"
            for turn in recent_turns
        ]
        sections.append("Recent turns:\n" + "\n".join(lines))
    text = "\n\n".join(sections).strip()
    return _trim_tail(text, max(300, int(max_chars)))


def append_exchange(
    *,
    recent_turns: list[dict[str, str]],
    user_turn_id: str,
    user_text: str,
    assistant_turn_id: str,
    assistant_text: str,
) -> list[dict[str, str]]:
    out = list(_normalize_turns(recent_turns))
    out.append(_turn("user", user_turn_id, user_text))
    if assistant_text.strip():
        out.append(_turn("assistant", assistant_turn_id, assistant_text))
    return out


def extract_edit_turn_id(edit: Any) -> str | None:
    if not isinstance(edit, dict):
        return None
    raw = edit.get("turn_id")
    if not isinstance(raw, str):
        return None
    value = raw.strip()
    return value or None


def response_message_text(resp: WorkflowRunResponse) -> str:
    if isinstance(resp.result, str) and resp.result.strip():
        return resp.result.strip()
    payload = resp.client_handler.get("payload") if isinstance(resp.client_handler, dict) else None
    if isinstance(payload, dict):
        for key in ("message", "summary", "prompt"):
            value = payload.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()
    if resp.errors and isinstance(resp.errors[0].message, str):
        return resp.errors[0].message.strip()
    return ""


def prepare_dialog_memory(
    *,
    req: WorkflowRunRequest,
    recent_messages_limit: int,
    summary_max_chars: int,
    context_max_chars: int,
) -> dict[str, Any]:
    summary, recent_turns, extra = memory_from_context(req.runtime.context if isinstance(req.runtime.context, dict) else None)
    summary, recent_turns = apply_edit_rewrite(
        summary=summary,
        recent_turns=recent_turns,
        edit_turn_id=extract_edit_turn_id(req.edit),
        recent_messages_limit=recent_messages_limit,
        summary_max_chars=summary_max_chars,
    )
    return {
        "summary": summary,
        "recent_turns": recent_turns,
        "context_extra": extra,
        "dialog_context": build_dialog_context(
            summary=summary,
            recent_turns=recent_turns,
            extra=extra,
            max_chars=context_max_chars,
        ),
    }


def _turn(role: str, turn_id: str, text: str) -> dict[str, str]:
    return {
        "role": "assistant" if role.strip().lower() == "assistant" else "user",
        "turn_id": turn_id.strip(),
        "text": text.strip(),
    }


def _normalize_turns(raw: Any) -> list[dict[str, str]]:
    if not isinstance(raw, list):
        return []
    out: list[dict[str, str]] = []
    for row in raw:
        if not isinstance(row, dict):
            continue
        role = str(row.get("role") or "").strip().lower()
        turn_id = str(row.get("turn_id") or "").strip()
        text = str(row.get("text") or "").strip()
        if role not in {"user", "assistant"} or not turn_id or not text:
            continue
        out.append({"role": role, "turn_id": turn_id, "text": _single_line(text, 500)})
    return out


def _merge_summary(current: str, overflow: list[dict[str, str]], summary_max_chars: int) -> str:
    rows = [f"{_role_label(row['role'])}: {_single_line(row['text'], 200)}" for row in overflow]
    merged = current.strip()
    chunk = " | ".join(rows).strip()
    if not chunk:
        return _trim_tail(merged, summary_max_chars)
    merged = f"{merged} | {chunk}".strip(" |")
    return _trim_tail(merged, summary_max_chars)


def _single_line(text: str, limit: int) -> str:
    one_line = " ".join(str(text or "").split())
    return one_line if len(one_line) <= limit else f"{one_line[: max(1, limit - 14)]}...<truncated>"


def _trim_tail(text: str, max_chars: int) -> str:
    clean = str(text or "").strip()
    if len(clean) <= max_chars:
        return clean
    return f"...{clean[-max_chars:]}"


def _role_label(role: str) -> str:
    return "User" if role == "user" else "Assistant"


def _artifact_context_block(extra: dict[str, Any] | None) -> str:
    if not isinstance(extra, dict):
        return ""
    artifact = extra.get(DEFAULT_ARTIFACT_CONTEXT_KEY)
    if not isinstance(artifact, dict):
        return ""

    sections: list[str] = []
    airports = artifact.get("last_airports")
    if isinstance(airports, list) and airports:
        lines: list[str] = []
        for row in airports[:5]:
            if not isinstance(row, dict):
                continue
            name = str(row.get("name") or "").strip()
            code = str(row.get("code") or "").strip()
            ident = f"{name} ({code})".strip() if name and code else (name or code)
            if not ident:
                continue
            status = str(row.get("status") or "").strip()
            lines.append(f"- {ident}{f', status={status}' if status else ''}")
        if lines:
            sections.append("Recent airports:\n" + "\n".join(lines))

    position = artifact.get("last_position")
    if isinstance(position, dict):
        city = str(position.get("city") or "").strip()
        lat = position.get("lat")
        lon = position.get("lon")
        parts: list[str] = []
        if city:
            parts.append(f"city={city}")
        if lat is not None and lon is not None:
            parts.append(f"coords=({lat},{lon})")
        if parts:
            sections.append("Recent position: " + ", ".join(parts))

    route = artifact.get("last_route")
    if isinstance(route, dict):
        distance = route.get("distance_km")
        if distance is not None:
            sections.append(f"Recent route distance_km={distance}")

    return "\n".join(sections).strip()

File: src_langgraph_rb_node/lib/src_langgraph_rb_node/runtime/memory.rb
# frozen_string_literal: true

module SrcLanggraphRbNode
  module Runtime
    module Memory
      DEFAULT_MEMORY_CONTEXT_KEY = "dialog_memory"
      DEFAULT_ARTIFACT_CONTEXT_KEY = "artifact_memory"
      module_function

      def memory_from_context(context)
        base = context.is_a?(Hash) ? Runtime::Util.extract_hash(context) : {}
        extra = base.each_with_object({}) do |(key, value), out|
          out[key] = value unless key.to_s == DEFAULT_MEMORY_CONTEXT_KEY
        end
        payload = Runtime::Util.extract_hash(base[DEFAULT_MEMORY_CONTEXT_KEY.to_sym] || base[DEFAULT_MEMORY_CONTEXT_KEY])
        summary = (payload[:summary] || "").to_s.strip
        recent = normalize_turns(payload[:recent_turns])
        [summary, recent, extra]
      end

      def context_with_memory(summary:, recent_turns:, extra:)
        out = Runtime::Util.extract_hash(extra)
        out[DEFAULT_MEMORY_CONTEXT_KEY.to_sym] = {
          summary: summary.to_s.strip,
          recent_turns: normalize_turns(recent_turns)
        }
        out
      end

      def apply_edit_rewrite(summary:, recent_turns:, edit_turn_id:, recent_messages_limit:, summary_max_chars:)
        return compact_memory(summary:, recent_turns:, recent_messages_limit:, summary_max_chars:) unless edit_turn_id

        idx = recent_turns.index { |turn| turn[:role] == "user" && turn[:turn_id] == edit_turn_id }
        return ["", []] unless idx

        compact_memory(
          summary: "",
          recent_turns: recent_turns[0...idx],
          recent_messages_limit: recent_messages_limit,
          summary_max_chars: summary_max_chars
        )
      end

      def compact_memory(summary:, recent_turns:, recent_messages_limit:, summary_max_chars:)
        normalized = normalize_turns(recent_turns)
        safe_limit = [2, recent_messages_limit.to_i].max
        safe_summary_max = [200, summary_max_chars.to_i].max
        return [trim_tail(summary, safe_summary_max), normalized] if normalized.length <= safe_limit

        overflow = normalized[0...-safe_limit]
        kept = normalized[-safe_limit..]
        [merge_summary(summary, overflow, safe_summary_max), kept]
      end

      def build_dialog_context(summary:, recent_turns:, extra:, max_chars:)
        sections = []
        artifact_block = artifact_context_block(extra)
        sections << artifact_block unless artifact_block.empty?
        sections << "Summary:\n#{summary.strip}" unless summary.to_s.strip.empty?
        unless recent_turns.empty?
          lines = recent_turns.map { |turn| "- #{role_label(turn[:role])}: #{turn[:text]}" }
          sections << "Recent turns:\n#{lines.join("\n")}"
        end
        trim_tail(sections.join("\n\n").strip, [300, max_chars.to_i].max)
      end

      def append_exchange(recent_turns:, user_turn_id:, user_text:, assistant_turn_id:, assistant_text:)
        out = normalize_turns(recent_turns)
        out << turn("user", user_turn_id, user_text)
        out << turn("assistant", assistant_turn_id, assistant_text) unless assistant_text.to_s.strip.empty?
        out
      end

      def extract_edit_turn_id(edit)
        Runtime::Util.text(Runtime::Util.extract_hash(edit)[:turn_id])
      end

      def response_message_text(resp)
        result = resp[:result]
        return result.strip if result.is_a?(String) && !result.strip.empty?

        payload = Runtime::Util.extract_hash(Runtime::Util.extract_hash(resp[:client_handler])[:payload])
        %i[message summary prompt].each do |key|
          value = payload[key]
          return value.strip if value.is_a?(String) && !value.strip.empty?
        end

        errors = Array(resp[:errors])
        first = Runtime::Util.extract_hash(errors.first)
        message = first[:message]
        message.is_a?(String) ? message.strip : ""
      end

      def prepare_dialog_memory(req:, recent_messages_limit:, summary_max_chars:, context_max_chars:)
        summary, recent_turns, extra = memory_from_context(req.dig(:runtime, :context))
        edit_turn_id = extract_edit_turn_id(req[:edit])
        summary, recent_turns = apply_edit_rewrite(
          summary: summary,
          recent_turns: recent_turns,
          edit_turn_id: edit_turn_id,
          recent_messages_limit: recent_messages_limit,
          summary_max_chars: summary_max_chars
        )
        extra.delete(DEFAULT_ARTIFACT_CONTEXT_KEY.to_sym) if edit_turn_id

        {
          summary: summary,
          recent_turns: recent_turns,
          context_extra: extra,
          dialog_context: build_dialog_context(
            summary: summary,
            recent_turns: recent_turns,
            extra: extra,
            max_chars: context_max_chars
          )
        }
      end

      def turn(role, turn_id, text)
        {
          role: role.to_s.strip.downcase == "assistant" ? "assistant" : "user",
          turn_id: turn_id.to_s.strip,
          text: text.to_s.strip
        }
      end
      private_class_method :turn

      def normalize_turns(raw)
        return [] unless raw.is_a?(Array)

        raw.filter_map do |row|
          next unless row.is_a?(Hash)

          item = Runtime::Util.extract_hash(row)
          role = item[:role].to_s.strip.downcase
          turn_id = item[:turn_id].to_s.strip
          text = item[:text].to_s.strip
          next unless %w[user assistant].include?(role) && !turn_id.empty? && !text.empty?

          { role: role, turn_id: turn_id, text: single_line(text, 1000) }
        end
      end
      private_class_method :normalize_turns

      def merge_summary(current, overflow, summary_max_chars)
        rows = overflow.map { |row| "#{role_label(row[:role])}: #{single_line(row[:text], 400)}" }
        merged = current.to_s.strip
        chunk = rows.join(" | ").strip
        return trim_tail(merged, summary_max_chars) if chunk.empty?

        trim_tail([merged, chunk].reject(&:empty?).join(" | "), summary_max_chars)
      end
      private_class_method :merge_summary

      def single_line(text, limit)
        one_line = text.to_s.split.join(" ")
        return one_line if one_line.length <= limit

        "#{one_line[0, [1, limit - 14].max]}...<truncated>"
      end
      private_class_method :single_line

      def trim_tail(text, max_chars)
        clean = text.to_s.strip
        return clean if clean.length <= max_chars

        "...#{clean[-max_chars..]}"
      end
      private_class_method :trim_tail

      def role_label(role)
        role == "user" ? "User" : "Assistant"
      end
      private_class_method :role_label

      def artifact_context_block(extra)
        return "" unless extra.is_a?(Hash)

        artifact = Runtime::Util.extract_hash(extra[DEFAULT_ARTIFACT_CONTEXT_KEY.to_sym] || extra[DEFAULT_ARTIFACT_CONTEXT_KEY])
        return "" if artifact.empty?

        lines = artifact.filter_map do |key, value|
          summary = artifact_value_summary(value)
          "- #{key}: #{summary}" unless summary.nil? || summary.empty?
        end
        return "" if lines.empty?

        "Context artifacts:\n#{lines.join("\n")}"
      end
      private_class_method :artifact_context_block

      def artifact_value_summary(value)
        if value.is_a?(Array)
          return "[]" if value.empty?

          items = value.first(5).filter_map { |row| artifact_value_brief(row) }
          return "list[#{value.length}]" if items.empty?

          suffix = value.length > 5 ? "; ..." : ""
          return "#{items.join('; ')}#{suffix}"
        end

        artifact_value_brief(value)
      end
      private_class_method :artifact_value_summary

      def artifact_value_brief(value)
        if value.is_a?(Hash)
          payload = Runtime::Util.extract_hash(value)
          label = artifact_entity_label(payload)
          return label unless label.empty?

          pairs = []
          payload.each do |key, item|
            scalar = artifact_scalar(item)
            next unless scalar

            pairs << "#{key}=#{scalar}"
            break if pairs.length >= 4
          end
          return pairs.join(", ") unless pairs.empty?

          return "object"
        end

        artifact_scalar(value).to_s
      end
      private_class_method :artifact_value_brief

      def artifact_entity_label(payload)
        label = payload[:label].to_s.strip
        label = payload[:name].to_s.strip if label.empty?
        label = payload[:title].to_s.strip if label.empty?
        code = payload[:code].to_s.strip
        ident = payload[:id].to_s.strip
        ident = payload[:key].to_s.strip if ident.empty?

        return "#{label} (#{code})" unless label.empty? || code.empty?
        return "#{label} (#{ident})" unless label.empty? || ident.empty? || ident == label
        return label unless label.empty?
        return code unless code.empty?
        return ident unless ident.empty?

        ""
      end
      private_class_method :artifact_entity_label

      def artifact_scalar(value)
        case value
        when nil
          nil
        when true
          "true"
        when false
          "false"
        when Numeric
          value.to_s
        when String
          clean = single_line(value, 80)
          clean.empty? ? nil : clean
        end
      end
      private_class_method :artifact_scalar
    end
  end
end
